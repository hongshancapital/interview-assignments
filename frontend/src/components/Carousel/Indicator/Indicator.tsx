import React from 'react';

import { State } from '../helper';
import { IndicatorItem, ProgressBar } from './styles';

interface IndicatorProps {
  index: number;
  state: State;
  duration: number;
  onClick: (index: number) => void;
}

export const Indicator = (props: IndicatorProps): JSX.Element => {
  const { index, state, duration, onClick } = props;

  const isCurrentIndicator = state === 'current';

  return (
    <IndicatorItem data-testid="indicator-item" onClick={() => onClick(index)}>
      {isCurrentIndicator && (
        <ProgressBar duration={duration} data-testid="progress-bar" />
      )}
    </IndicatorItem>
  );
};
