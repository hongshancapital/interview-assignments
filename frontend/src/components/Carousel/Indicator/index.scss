.indicator-container {
    .indicator-item {
        display: inline-flex;
        width: 50px;
        height: 2px;
        background-color: #a9a9a9;
        cursor: pointer;
        margin-right: 10px;
        border-radius: 1px;
        overflow: hidden;
    }
    .indicator-fill {
        width: 0;
        height: 100%;
        background-color: #fafafa;
    }
}
