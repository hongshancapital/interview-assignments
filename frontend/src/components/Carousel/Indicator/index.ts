export { Indicator } from './Indicator';
export { IndicatorContainer } from './styles';
