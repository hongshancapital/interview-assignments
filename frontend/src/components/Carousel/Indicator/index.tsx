import React, { CSSProperties, useCallback, useMemo } from 'react';
import './index.scss';

export interface IndicatorProps {
    /**当前索引 */
    curIndex: number;
    /**总数 */
    total: number;
    /**切换间隔(ms) */
    interval: number;
    /**点击切换页面回调 */
    onChange?: (index: number) => void;
}

export default function Indicator(props: IndicatorProps) {
    const { curIndex, total, interval, onChange = _ => _ } = props;
    const items = useMemo(
        () =>
            Array(total).fill(0).map((_, i) => {
                if (i === Number(curIndex)) {
                    return { width: '100%', transition: `width ${interval}ms linear` } as CSSProperties;
                }
                return {};
            }),
        [total, curIndex, interval]
    );
    const handleSwitch = useCallback((e) => {
        const { index } = e?.target?.dataset || {};
        if (index === undefined) return;
        onChange(index);
    }, [onChange]);

    return (
        <div className='indicator-container' onClick={handleSwitch}>
            {items.map((it, i) => (
                <div className='indicator-item' key={i} data-index={i}>
                    <span className='indicator-fill' style={it} />
                </div>
            ))}
        </div>
    );
}
