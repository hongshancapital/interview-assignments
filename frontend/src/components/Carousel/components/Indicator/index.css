.indecator-wrapper{
  position: absolute;
  z-index: 1;
  bottom:50px;
  margin-left: 50%;
  transform: translateX(-50%);
  width : 160px;
  height: 10px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.indecator-wrapper .item{
  display: flex;
  width:48px;
  height: 2px;
  background: #ccc;
  margin: 0 4px;
  border-radius: 1px;
}

.indecator-wrapper .item .inner-color{
  display: flex;
  height: 100%;
  background: white;
}
