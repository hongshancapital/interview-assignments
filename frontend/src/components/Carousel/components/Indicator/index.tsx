import React from "react";
import "./index.css";
export interface IndecatorProps {
  length: number;
  delay: number;
  index: number;
}
export const Indecator = (props: IndecatorProps) => {
  const { index: currentItemIndex, length, delay } = props;
  const itemList = Array.from({ length })
    .fill(0)
    .map((_, index) => index);

  return (
    <div className="indecator-wrapper">
      {itemList.map((item) => (
        <div key={item} className={"item"}>
          <div
            style={{
              transition: item === currentItemIndex ? `${delay}ms` : "0ms",
              width: item === currentItemIndex ? "100%" : "0%",
            }}
            className="inner-color"
          ></div>
        </div>
      ))}
    </div>
  );
};

export default Indecator;
