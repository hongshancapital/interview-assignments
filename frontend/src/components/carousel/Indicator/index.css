.indicator-control {
  width: 56px;
  height: 3px;
  border-radius: 3px;
  margin: 0 5px;
  background-color: #a0a0a0;
  position: relative;
  display: inline-block;
  overflow: hidden;
}

.indicator-control .indicator-bar {
  width: 100%;
  height: 100%;
}
