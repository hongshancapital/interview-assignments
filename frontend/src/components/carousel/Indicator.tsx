import './Indicator.css';

export interface IndicatorsProps {
  currentIndex: number;
  total: number;
  duration: number;
}

const Indicator = ({ currentIndex, total, duration }: IndicatorsProps) => {
  return (
    <ul className={'indicator'}>
      {new Array(Math.max(0, total)).fill(null).map((_, index) => {
        return (
          <li
            key={index}
            className={[
              'indicator__item',
              currentIndex === index ? 'indicator__item_active' : '',
            ]
              .filter(Boolean)
              .join(' ')}
          >
            <span
              className={'indicator__progress'}
              style={
                currentIndex === index
                  ? { animationDuration: `${duration}ms` }
                  : {}
              }
            />
          </li>
        );
      })}
    </ul>
  );
};
Indicator.displayName = 'Indicator';

export default Indicator;
