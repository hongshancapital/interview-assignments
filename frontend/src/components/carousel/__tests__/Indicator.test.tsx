import { render } from '@testing-library/react';
import Indicator from '../Indicator';

describe('Indicator', () => {
  it('should render Indicator component', () => {
    const { container, asFragment } = render(<Indicator currentIndex={0} total={3} duration={1000} />);
    expect(asFragment()).toMatchSnapshot();
    expect(container.querySelectorAll('.indicator__item')[0]).toHaveClass('indicator__item_active');
  });

  it('should shows the progress when indicator is active', () => {
    const { container, rerender } = render(<Indicator currentIndex={0} total={3} duration={1000} />);
    expect(container.querySelectorAll('.indicator__item')[0]).toHaveClass('indicator__item_active');
    expect(container.querySelectorAll('.indicator__item .indicator__progress')[0]).toHaveStyle({ animationDuration: `1000ms` });
    rerender(<Indicator currentIndex={1} total={3} duration={1000} />);
    expect(container.querySelectorAll('.indicator__item')[1]).toHaveClass('indicator__item_active');
  });
});
