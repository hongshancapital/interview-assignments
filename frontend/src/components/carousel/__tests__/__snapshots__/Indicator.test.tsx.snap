// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indicator should render Indicator component 1`] = `
<DocumentFragment>
  <ul
    class="indicator"
  >
    <li
      class="indicator__item indicator__item_active"
    >
      <span
        class="indicator__progress"
        style="animation-duration: 1000ms;"
      />
    </li>
    <li
      class="indicator__item"
    >
      <span
        class="indicator__progress"
      />
    </li>
    <li
      class="indicator__item"
    >
      <span
        class="indicator__progress"
      />
    </li>
  </ul>
</DocumentFragment>
`;
